core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorPageSource.java

```diff
@@ -31,6 +32,16 @@
      */
     long getCompletedBytes();
 
+    /**
+     * Gets the number of input rows processed by this page source so far.
+     * By default, the positions count of the page returned from getNextPage
+     * is used to calculate the number of input rows.
+     */
+    default OptionalLong getCompletedPositions()
```

Comment:
There’s an extensible getMetrics that could be used to return this.

Reply:
I am also using this in the table scan operators to populate values for physical input positions.
I think we'd want to use an explicit API for that instead of getMetrics.

Reply:
Ok, that's fine for now. I think eventually we'll want to migrate as many metrics as possible to the new API.